scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +444,238 @@
-          (cond
-            (empty? matches) (decline-offers driver (:ids offer))
-            (not (compare-and-set! pending-jobs scheduler-contents new-scheduler-contents))
-            (do (log/info "Pending job atom contention encountered")
+(defn ids-of-backfilled-instances
+  "Returns a list of the ids of running, backfilled instances of the given job"
+  [job]
+  (->> (:job/instance job)
+       (filter #(and (= :instance.status/running (:instance/status %))
+                     (:instance/backfilled? %)))
+       (map :db/id)))
+
+(defn process-matches-for-backfill
+  "This computes some sets:
+   
+   fully-processed: this is a set of job uuids that should be removed from the list scheduler jobs. These are not backfilled
+   upgrade-backfill: this is a set of instance datomic ids that should be upgraded to non-backfill
+   backfill-jobs: this is a set of job uuids for jobs that were matched, but are in the backfill QoS class
+   "
+  [scheduler-contents head-of-considerable matched-jobs]
+  (let [matched-job-uuids (set (mapv :job/uuid matched-jobs))
+        head-uuid (:job/uuid head-of-considerable)]
+    (loop [scheduler-contents scheduler-contents
+           backfilling? false
+           matched-head? false
+           fully-processed #{}
+           upgrade-backfill #{}
+           backfill-jobs #{}]
+      (if (seq scheduler-contents)
+        (let [scheduler-contents' (rest scheduler-contents)
+              pending-job (first scheduler-contents)
+              pending-job-uuid (:job/uuid pending-job)
+              pending=head? (= pending-job-uuid head-uuid)]
+          (if (contains? matched-job-uuids pending-job-uuid)
+            (if backfilling?
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill (conj backfill-jobs pending-job-uuid))
+              (recur scheduler-contents' false (or pending=head? matched-head?) (conj fully-processed pending-job-uuid) upgrade-backfill backfill-jobs))
+            (if-let [backfilled-ids (and (not backfilling?) (seq (ids-of-backfilled-instances pending-job)))]
+              (recur scheduler-contents' false (or pending=head? matched-head?) fully-processed (into upgrade-backfill backfilled-ids) backfill-jobs)
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill backfill-jobs))))
+        {:fully-processed fully-processed
+         :upgrade-backfill upgrade-backfill
+         :backfill-jobs backfill-jobs
+         :matched-head? matched-head?}))))
+
+(defn handle-resource-offers!
+  "Gets a list of offers from mesos. Decides what to do with them all--they should all
+   be accepted or rejected at the end of the function."
+  [conn driver ^TaskScheduler fenzo fid pending-jobs num-considerable offers-chan offers]
+  (log/debug "invoked handle-resource-offers!")
+  (let [offer-stash (atom nil)] ;; This is a way to ensure we never lose offers fenzo assigned if an errors occures in the middle of processing
```

Comment:
The issue (and reason I believe an atom was used) is that a Try Block introduces its own lexical scope.  If we want the exception handling of the try block to reset the offer stash with a value that is set inside the try block, an atom is the simplest way to do that, since the atom (unlike vars) can acquire a value that is available outside of the scope where it is set.

I believe We could avoid using an atom, but we would have to rearrange this whole function significantly, and partially change the way exceptions are handled in order to do so.  We could have the parts that precede (reset! offer-stash) in one try block (maybe even a separate function), that returns a tuple of  "considerable" and "matches" ...  then reset the offer stash, then start a new try block that does the rest of what the existing function does (and handles exceptions in the same way).